src/lib.rs

```diff
@@ -1034,20 +1035,80 @@ impl<T: Clone + Integer + Signed> Signed for Ratio<T> {
 }
 
 // String conversions
-impl<T> fmt::Display for Ratio<T>
-where
-    T: fmt::Display + Eq + One,
-{
-    /// Renders as `numer/denom`. If denom=1, renders as numer.
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        if self.denom.is_one() {
-            write!(f, "{}", self.numer)
-        } else {
-            write!(f, "{}/{}", self.numer, self.denom)
+macro_rules! impl_formatting {
+    ($fmt_trait:ident, $prefix:expr, $fmt_str:expr, $fmt_alt:expr) => {
+        impl<T: $fmt_trait + Clone + Integer> $fmt_trait for Ratio<T> {
+            #[cfg(feature = "std")]
+            fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+                let pre_pad = if self.denom.is_one() {
+                    format!($fmt_str, self.numer)
+                } else {
+                    if f.alternate() {
+                        format!(concat!($fmt_str, "/", $fmt_alt), self.numer, self.denom)
+                    } else {
+                        format!(concat!($fmt_str, "/", $fmt_str), self.numer, self.denom)
+                    }
+                };
+                if f.sign_plus() {
+                    let pre_pad = pre_pad.trim_start_matches('-');
+                    let non_negative = self.numer >= T::zero();
+                    f.pad_integral(non_negative, $prefix, pre_pad)
+                } else {
+                    f.pad_integral(true, $prefix, &pre_pad)
+                }
+            }
+            #[cfg(not(feature = "std"))]
+            fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+                if f.sign_plus() && self.numer >= T::zero() {
```

Comment:
I think we could fold some of this duplication by passing `format_args!` to a nested function taking `fmt::Arguments` -- see [how `num-complex` does this](https://github.com/rust-num/num-complex/blob/7fb4f1c0f0f0d293c18df8d5b2b03362605b3862/src/lib.rs#L1123). So on the outer part you'd format the str/alt/denom variations, then the inner part would write that to `f` with or without the `+` prefix.

(Don't take `num-complex` as a paragon -- I don't think that implementation has solved all the issues you're tackling here.)

Reply:
Or, maybe just set the "+" as a dynamic prefix, else "", written with an extra `{}` at the beginning.